src/oop/object_interactions/medium/q4/MainTest.java

```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of('A', 10.3, 'B', 4.9),
+                Arguments.of('A', 15.5, 'A', 7.2),
+                Arguments.of('A', 6.4, 'A', minBlood),
+                Arguments.of('A', 0, 'A', minBlood - 1),
+                Arguments.of('A', 0.5, 'A', minBlood + 1),
+                Arguments.of('B', 0, 'A', 2.5),
+                Arguments.of('O', 3.4, 'B', 5.7)
+        );
+    }
+
+    @Test
+    public void transfusionFromIsDefinedCorrectly() {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(transfusionFromMethodName, new Class<?>[]{Snail.class}), String.format(incorrectMethodDefinition, transfusionFromMethodName, className));
+        assertTrue(testClass.hasModifier(transfusionFromMethodName, new Class<?>[]{Snail.class}, "public"), String.format(incorrectModifierMessage, transfusionFromMethodName, className));
+        assertTrue(testClass.hasReturnType(transfusionFromMethodName, new Class<?>[]{Snail.class}, Void.TYPE), String.format(incorrectReturnType, transfusionFromMethodName, className));
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependent1")
+    public void correctTransfusionFromMethod(char type, double volume, char donType, double donVolume) throws Throwable {
+        Object[][] arguments = {
+                {type, char.class},
+                {volume, double.class}
+        };
+        Snail donor = new Snail(donType, donVolume);
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(transfusionFromMethodName, new Class<?>[]{Snail.class}), String.format(incorrectMethodDefinition, transfusionFromMethodName, className));
+        assertTrue(testClass.hasModifier(transfusionFromMethodName, new Class<?>[]{Snail.class}, "public"), String.format(incorrectModifierMessage, transfusionFromMethodName, className));
+        assertTrue(testClass.hasReturnType(transfusionFromMethodName, new Class<?>[]{Snail.class}, Void.TYPE), String.format(incorrectReturnType, transfusionFromMethodName, className));
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {donor, Snail.class}
```

Comment:
Also, please do not call the class like this in case they modify something, which will cause out code not to compile. There is a way to get the type using `ObjectTest`